pkg/quota/image/imagestreamimport_evaluator.go

```diff
@@ -0,0 +1,220 @@
+// ImageStreamImport kind.
+func (c *imageStreamImportUsageComputer) Usage(object runtime.Object) kapi.ResourceList {
+	isi, ok := object.(*imageapi.ImageStreamImport)
+	if !ok {
+		return kapi.ResourceList{}
+	}
+
+	usage := map[kapi.ResourceName]resource.Quantity{
+		imageapi.ResourceImageStreamTags:   *resource.NewQuantity(0, resource.DecimalSI),
+		imageapi.ResourceImageStreamImages: *resource.NewQuantity(0, resource.DecimalSI),
+	}
+
+	if !isi.Spec.Import || (len(isi.Spec.Images) == 0 && isi.Spec.Repository == nil) {
+		return usage
+	}
+
+	// collect all the image references already tagged in a project
+	iss, err := c.osClient.ImageStreams(isi.Namespace).List(kapi.ListOptions{})
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to list image streams: %v", err))
+		return usage
+	}
+
+	for _, imageStream := range iss.Items {
+		c.ProcessImageStreamImages(&imageStream, false, func(ref string, inSpec, _ bool) error {
+			if !c.processedSpecRefs.Has(ref) && inSpec {
+				c.processedSpecRefs.Insert(ref)
+			}
+			if !c.processedRefs.Has(ref) {
+				c.processedRefs.Insert(ref)
+			}
+			return nil
+		})
+	}
+
+	// first process and remember individual images
+	imagesUsage := c.getImagesUsageIncrement(isi)
+	// second process a repository whose tags shall we import
+	repositoryUsage := c.getRepositoryUsageIncrement(isi)
+
+	return kquota.Add(imagesUsage, repositoryUsage)
+}
+
+func (c *imageStreamImportUsageComputer) getImagesUsageIncrement(isi *imageapi.ImageStreamImport) kapi.ResourceList {
+	specRefsIncrement := resource.NewQuantity(0, resource.DecimalSI)
+	allRefsIncrement := resource.NewQuantity(0, resource.DecimalSI)
+
+	for _, imageSpec := range isi.Spec.Images {
+		if imageSpec.From.Kind != "DockerImage" {
+			continue
+		}
+		ref, err := c.GetImageReferenceForObjectReference(isi.Namespace, &imageSpec.From)
+		if err != nil {
+
+		}
+		if !c.processedSpecRefs.Has(ref) {
+			c.processedSpecRefs.Insert(ref)
+			specRefsIncrement.Set(specRefsIncrement.Value() + 1)
+		}
+		if !c.processedRefs.Has(ref) {
+			c.processedRefs.Insert(ref)
+			allRefsIncrement.Set(allRefsIncrement.Value() + 1)
+		}
+	}
+
+	return map[kapi.ResourceName]resource.Quantity{
+		imageapi.ResourceImageStreamTags:   *specRefsIncrement,
+		imageapi.ResourceImageStreamImages: *allRefsIncrement,
+	}
+}
+
+// getRepositoryUsageIncrement returns a usage increment for ImageStreamImport.Spec.Repository. The repository
+// causes image importer to import all the repository tags in an image stream. We don't attempt to resolve
+// number of tags here. We just expect the worst case scenario where the max limit of import tags per
+// repository is reached and decrease it by the number of image references pointing to the repository that are
+// already tagged in a project.
+func (c *imageStreamImportUsageComputer) getRepositoryUsageIncrement(isi *imageapi.ImageStreamImport) kapi.ResourceList {
```

Comment:
To make it precise, it still evaluates other image streams in the project so that only new references are counted.